apollo-federation/src/sources/connect/validation/connect/errors.rs

```diff
@@ -0,0 +1,331 @@
+//! Parsing and validation for `@connect(errors:)`
+
+use std::fmt::Display;
+
+use apollo_compiler::Name;
+use apollo_compiler::Node;
+use apollo_compiler::ast::Value;
+use multi_try::MultiTry;
+use shape::Shape;
+
+use crate::sources::connect::JSONSelection;
+use crate::sources::connect::Namespace;
+use crate::sources::connect::spec::schema::ERRORS_ARGUMENT_NAME;
+use crate::sources::connect::spec::schema::ERRORS_EXTENSIONS_ARGUMENT_NAME;
+use crate::sources::connect::spec::schema::ERRORS_MESSAGE_ARGUMENT_NAME;
+use crate::sources::connect::string_template::Expression;
+use crate::sources::connect::validation::Code;
+use crate::sources::connect::validation::Message;
+use crate::sources::connect::validation::coordinates::ConnectDirectiveCoordinate;
+use crate::sources::connect::validation::expression;
+use crate::sources::connect::validation::expression::Context;
+use crate::sources::connect::validation::graphql::GraphQLString;
+use crate::sources::connect::validation::graphql::SchemaInfo;
+
+/// A valid, parsed (but not type-checked) `@connect(errors:)`.
+pub(super) struct Errors<'schema> {
+    message: Option<ErrorsMessage<'schema>>,
+    extensions: Option<ErrorsExtensions<'schema>>,
+}
+
+impl<'schema> Errors<'schema> {
+    /// Parse the `@connect(errors:)` argument and run just enough checks to be able to use the
+    /// argument at runtime. More advanced checks are done in [`Self::type_check`].
+    ///
+    /// Two sub-pieces are always parsed, and the errors from _all_ of those pieces are returned
+    /// together in the event of failure:
+    /// 1. `errors.message` with [`ErrorsMessage::parse`]
+    /// 2. `errors.extensions` with [`ErrorsExtensions::parse`]
+    ///
+    /// The order these pieces run in doesn't matter and shouldn't affect the output.
+    pub(super) fn parse(
+        coordinate: ConnectDirectiveCoordinate<'schema>,
+        schema: &'schema SchemaInfo,
+    ) -> Result<Self, Vec<Message>> {
+        let Some((errors_arg, _errors_arg_node)) = coordinate
+            .directive
+            .specified_argument_by_name(&ERRORS_ARGUMENT_NAME)
+            .and_then(|arg| Some((arg.as_object()?, arg)))
+        else {
+            return Ok(Self {
+                message: None,
+                extensions: None,
+            });
+        };
+
+        ErrorsMessage::parse(errors_arg, coordinate, schema)
+            .map_err(|err| vec![err])
+            .and_try(
+                ErrorsExtensions::parse(errors_arg, coordinate, schema).map_err(|err| vec![err]),
+            )
+            .map_err(|nested| nested.into_iter().flatten().collect())
+            .map(|(message, extensions)| Self {
+                message,
+                extensions,
+            })
```

Comment:
Should be able to do this:

```suggestion
        ErrorsMessage::parse(errors_arg, coordinate, schema)
            .and_try(ErrorsExtensions::parse(errors_arg, coordinate, schema))
            .map(|(message, extensions)| Self {
                message,
                extensions,
            })
```

The only reason we do the Vec messiness elsewhere is because _some_ of the things we're trying return multiple errors, and some return a single error.

Reply:
Good call. Updated!